### PR TITLE
MF-426 - Add optional MF_CA_CERTS env variable to allow GRPC client to use TLS certs

### PR DESCRIPTION
**Pull request by nwest1**
### What does this do?
Allows a user to specify a path to their own trusted certs file via environmental variable `MF_CA_CERTS`. 
### Which issue(s) does this PR fix/relate to?
related #426 
### List any changes that modify/break current functionality
None
### Have you included tests for your changes?
Not yet
### Did you document any new/modified functionality?
Not yet
### Notes
Wanted to get the idea out for feedback before moving onto other components

### CODE REVIEW

**Comment by drasko** [comment]
@nwest1 this PR looks good to me.

You will need to sign-off the commit in order for DCO check to pass

Just a question for the reference - is there some special way to deploy `things` service in order for it to provide TLS connection (private/public cert)?

**Comment by nwest1** [comment]
@drasko The SSL termination is currently handled by the ingress and not directly by the service itself. I can add that capability as discussed. Would you prefer all in one PR or leave this one for clients?

**Comment by drasko** [comment]
@nwest1 let's leave this one, making the PRs smaller.

Also, would WS and MQTT adapter be needing similar functionality? I guess yes, but if you are not using them or not capable to test them, then someone else from Mainflux core team can do these.

**Comment by nwest1** [comment]
Yes, MQTT will need to be changed in a similar fashion - I'll work on that soon. Though it looks like similar changes would need to happen in the WS adapter, we're not using it and aren't setup to test it.

File: cmd/things/main.go
Comment (dborovcanin) [inline cmd/things/main.go]: Can you, please, update the docs and docker-compose as well (README.md)?

File: cmd/http/main.go
Comment (dborovcanin) [inline cmd/http/main.go]: If `url.Parse(cfg.ThingsURL)` fails, `thingsURL` is possible an empty value, or nil, in case of the pointer, so calling `thingsURL.Host` is dangerous.

File: cmd/http/main.go
Comment (dborovcanin) [inline cmd/http/main.go]: Does it make sense to proceed in case of failure (maybe `os.Exit(1)` here)?

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: cmd/http/main.go
Comment (chombium) [inline cmd/http/main.go]: @nwest1 I think, It would be good if in case that no CA certs are set, to log a warning so that the user knows that the gRPC communication is not encrypted.

**Review by chombium — CHANGES_REQUESTED** [review]
An update of the documentation with explanation about how this feature works and how to configure it is also needed

File: cmd/things/main.go
Comment (nwest1) [reply cmd/things/main.go]: I think I've covered where this makes sense - should I leave the docker-compose with the default (blank) or do you want to include a sample CA along with your existing self-signed certs?

File: cmd/things/main.go
Comment (nwest1) [reply cmd/things/main.go:165]: I've removed the url parsing as well, as the `serverNameOverride` is actually for testing: https://godoc.org/google.golang.org/grpc/credentials#NewClientTLSFromCert

File: cmd/http/main.go
Comment (nwest1) [reply cmd/http/main.go]: Updated!

File: cmd/http/main.go
Comment (manuio) [inline cmd/http/main.go]: @nwest1 I would put this in a separate function as you did here: https://github.com/mainflux/mainflux/blob/897d4373b15b5fbd7b346086a3552929c3fedef4/cmd/things/main.go#L156

**Review by manuio — CHANGES_REQUESTED** [review]

File: cmd/http/main.go
Comment (chombium) [reply cmd/http/main.go]: @nwest1 cool, but I personally think logger.Warn might be more appropriate.

@drasko, @dusanb94 what do you think about this?

Comment (nwest1) [reply cmd/http/main.go]: All the other `logger.Warn`s are currently related to some error happening, but the app will continue to run. Though I think security is important, it's not in error, which is why I chose info. 

Happy to change this to whatever makes sense for the project.

File: cmd/http/main.go
Comment (nwest1) [reply cmd/http/main.go]: Done!

File: cmd/http/main.go
Comment (anovakovic01) [reply cmd/http/main.go]: @nwest1 I agree with you, we should use warn only for errors. This is valid flow, so I think that we should leave info.

Comment (chombium) [reply cmd/http/main.go]: @nwest1 thanks for checking on other warning usages. I totally agree with you. Let's leave it as an info message.

File: cmd/things/main.go
Comment (chombium) [reply cmd/things/main.go]: @nwest1 you can reuse the certs and ca from [here](https://github.com/mainflux/mainflux/tree/master/docker/ssl). We use them for nginx for our dev environment .

Regarding the docs, you should add a subsection in the [getting started guide](https://github.com/mainflux/mainflux/blob/master/docs/getting-started.md) with a short description how to setup the encryption and what happens if the certificate configuration is missing ;)

**Comment by anovakovic01** [comment]
@nwest1 @drasko I think that we should add cert support for service providers too (`users` and `things`) in this PR.

File: cmd/things/main.go
Comment (nwest1) [reply cmd/things/main.go]: I updated the getting started guide with some simple explanation. Unless you'd like me to configure nginx for grpc load balancing, I'm not sure it makes sense to secure this connection.

Another option would be to add a separate ENV var to control the `grpc.WithInsecure()` rather than the current logic. That pattern might scale out better than controlling two options with one parameter.

Comment (chombium) [reply cmd/things/main.go]: @nwest1 it looks good, we can change that in future if there is a need it and if we notice some problems.

I would ask you one more thing and then we are done regarding the docs :)
Please add a note in the getting started guide that if the env vars for setting the ca are omitted, the gRPC communication will not be encrypted.

**Comment by nwest1** [comment]
@anovakovic01 @drasko I'm happy to add cert support in this PR.

**Comment by nwest1** [comment]
I'm testing out the docker-compose portion, but am running into certificate validation errors. They're being eaten at the service level (https://github.com/mainflux/mainflux/blob/master/things/service.go#L149) 

It's looking for a `users` and/or `things` host cert to be presented - where in the current configuration it's `localhost`.

Would you like me to configure this differently or would it be possible to generate two more cert+key combos for the two hostnames expected? The CA key is not available in the repo.

File: cmd/things/main.go
Comment (nwest1) [reply cmd/things/main.go]: Added :) Let me know if there are any other details!

File: http/README.md
Comment (nwest1) [inline http/README.md]: Should this be `MF_HTTP_ADAPTER_CA_CERTS`? Or should `MF_THINGS_CA_CERTS` be `MF_CA_CERTS`?

Comment (drasko) [reply http/README.md]: I think it should be `MF_THINGS_CA_CERTS`, as CA issues private/public cert pair for server - which in this case is `things`.

**Comment by nwest1** [comment]
SSL changes: using https://github.com/mainflux/proxy/blob/master/certs/generate-CA.sh I recreated the ca and server cert + key combo with `localhost` `users` and `things` as Subject Alternative Names.  

I'm not sure why github thinks there are conflicts in docker-compose as they are only additions, but maybe someone else can resolve.

**Comment by drasko** [comment]
@nwest1 thanks a lot, looks good - we'll take a look over the WE.

**Review by manuio — DISMISSED** [review]

**Comment by chombium** [comment]
Hi @nwest1,

thanks for your engagement and the fast responses to our comments. 
Please merge the latest changes from the master and resolve the conflicts.

Two questions out of curiosity:
1. Why does the `nginx` containter depends on `things` the `users` containers?
2. Why are the following certificate files changed:
*  docker/ssl/certs/ca.crt 
*  docker/ssl/certs/mainflux-server.crt
* docker/ssl/certs/mainflux-server.key

File: things/README.md
Comment (anovakovic01) [inline things/README.md:28]: Please add `MF_THINGS_SERVER_CERT` and `MF_THINGS_SERVER_KEY` to docs.

File: cmd/users/main.go
Comment (anovakovic01) [inline cmd/users/main.go:58]: Add `MF_USERS_SERVER_CERT` and `MF_USERS_SERVER_KEY` to `users/README.md`.

**Review by anovakovic01 — CHANGES_REQUESTED** [review]

**Review by dborovcanin — COMMENTED** [review]
@nwest1 Great job! Please check if docs are up to date (env variables as well as env var list when starting services) and update the branch, so that we can merge.

**Review by dborovcanin — DISMISSED** [review]

**Comment by nwest1** [comment]
@chombium 
1. I was running into a few errors when configuring, but this is not needed and I've removed.
2. I regenerated the SSL certificates, adding `users` `things` and `localhost` as subject alternative names, so the same certificate can be used by all servers + nginx.

Shall I include the CA key as well? I used https://github.com/mainflux/proxy/blob/master/certs/generate-CA.sh to generate these.

**Comment by nwest1** [comment]
Happy to contribute! This branch should be all fixed up at this point. 

Thanks for all the reviews and comments - I appreciate the quick feedback.

**Comment by drasko** [comment]
@nwest1 please update the branch, we just merged some other stuff so your PR is out of sync now :)

**Review by dborovcanin — DISMISSED** [review]

File: users/README.md
Comment (drasko) [inline users/README.md]: Why is port relevant here?

File: http/README.md
Comment (drasko) [inline http/README.md]: Please align tabs/spaces for readability.

**Review by drasko — CHANGES_REQUESTED** [review]
@nwest1 almost there - few minor tweaks left.

**Review by drasko — APPROVED** [review]
LGTM

**Review by anovakovic01 — APPROVED** [review]
LGTM!

**Review by dborovcanin — APPROVED** [review]

**Comment by drasko** [comment]
@chombium if OK with you we would like to proceed merging this one.

**Review by chombium — APPROVED** [review]

**Comment by chombium** [comment]
LGTM! Great work @nwest1

**Comment by drasko** [comment]
Merged! Thanks a lot @nwest1 - great contribution!

**Pull request merged**
